api/segments/serializers.py

```diff
@@ -81,11 +81,15 @@ def create(self, validated_data):
         self._update_or_create_metadata(metadata_data, segment=segment)
         return segment
 
-    def update(self, instance, validated_data):
+    def update(self, instance: Segment, validated_data: dict[str, typing.Any]) -> None:
         # use the initial data since we need the ids included to determine which to update & which to create
         rules_data = self.initial_data.pop("rules", [])
         metadata_data = validated_data.pop("metadata", [])
         self.validate_segment_rules_conditions_limit(rules_data)
+
+        # Create a version of the segment now that we're updating.
+        instance.deep_clone()
```

Comment:
We might need to be more clever here... I'm worried about failure modes in the update that would mean we end up with phantom versions. 

We could wrap the below in a `try` / `except` perhaps, or we could move this logic to the model.

Reply:
Do you mean a `try` / `except` in the logic that handles updates? If so, we'd be altering the segment before cloning it which would defeat the purpose, no?

Reply:
Well, yes, I guess in an ideal world, we'd create the clone without persisting it to the database before doing the update and then persist it once the update has happened successfully. 

Since that's not really possible, due to all the foreign key relationships, we could instead just handle the deletion (I guess we'd want to hard delete too) of the already created clone in the `except` - although this still isn't bulletproof.

Reply:
Ok I've implemented this. Sorta feels a bit cludgy but I guess it's the only way forward.